src/main/scala/scalaz/stream/async/TaskDriver.scala

```diff
@@ -0,0 +1,164 @@
+   *
+   * Whenever task is completed, following is guaranteed
+   * - if the interrupt was received while evaluating the task (pure code) then `cb` is called with result of the task evaluation
+   * - if the interrupt was received while evaluating the task (async code) then evaluated result is discarded
+   * - otherwise `cb` is called with result of task evaluatin
+   *
+   * @param task    Task to evaluate
+   * @param cb      cb to invoke after evalutation is done, or when task is interrupted
+   */
+  def runTaskAsyncInterruptibly[A](task: Task[A], cb: (Throwable \/ A) => Unit): () => Unit = {
+    sealed trait M
+    case object Interrupt extends M
+    case class Complete(r: Throwable \/ A) extends M
+    case class CompleteBind(r: Any, g: Any => Future[Throwable \/ A]) extends M
+
+    var open: Boolean = true
+
+    def complete(r: (Throwable \/ A)): Unit = {
+      open = false
+      cb(r)
+    }
+
+    def tryCompleteTask(current: Future[Throwable \/ A], actor: => Actor[M]): Unit = {
+      current.step match {
+        case Future.Now(r) => complete(r)
+        case Future.Async(onFinish) =>
+          onFinish((r: Throwable \/ A) => Trampoline.done(actor ! Complete(r)))
+        case Future.BindAsync(onFinish, g) =>
+          onFinish((r: Any) => Trampoline.done(actor ! CompleteBind(r, g)))
+
+      }
+    }
+
+    lazy val actor: Actor[M] = Actor[M]({
+      case Complete(r) if open => complete(r)
+      case CompleteBind(r, g) if open => tryCompleteTask(g(r), actor)
+      case Interrupt if open => complete(left(Task.TaskInterrupted))
+      case _ => //no-op
+
+    })(Strategy.Sequential)
+
+    tryCompleteTask(task.get, actor)
+    () => {actor ! Interrupt}
+  }
+
+
+  /**
+   * Asynchronous execution of this Process. Note that this method is not resource safe unless
+   * callback is called with _left_ side completed. In that case it is guaranteed that all cleanups
+   * has been successfully completed.
+   *
+   * User of this method is responsible for any cleanup actions to be performed by running the
+   * next Process obtained on right side of callback.
+   *
+   * This method returns a function, that when applied, causes the running computation to be interrupted.
+   * That is useful of process contains any asynchronous code, that may be left with incomplete callbacks.
+   * If the evaluation of the process is interrupted, then the interruption is only active if the callback
+   * was not completed before, otherwise interruption is no-op.
+   *
+   * This implementation also guarantees that when `cb` is invoked on right side, then sequence is always non-empty.
+   * This is usefully to run two processes truly in parallel w/o consulting higher level combinator (i.e. wye/merge).
+   *
+   * This implementation as well guaranteed that any onComplete // onFail // onKill code is consulted exactly once
+   * whether the computation was or was not interrupted.
```

Comment:
> You are right we may introduce the InterruptedA but won't that be too much work to resolve it everywhere in code? I mean awaitInterruptibly essentially does the same job, and we can live with same algebra?

Well, for me, `awaitInterruptibly` looks a heck of a lot like `awaitOr`.  I don't think it would be too much trouble to resolve `Interrupted` being added to the `EarlyCause` algebra, and since it allows the existing combinators to cover a slightly extended use-case while avoiding the need to introduce a nearly-redundant combinator, I think that it would be my preferred option if the strengthened `onComplete` was off the table.

> The problem I see there is that you will have to evaluate rcv twice (once on interrupt and eventually second time when a was produced on after-interrupt complete task, which I think was something we tried to avoid? Solution with awaitInterruptibly allows you to resolve this just in Task, and clearly define piece of code that will never be run twice.

`rcv` should be pure.  If it's not, then users have other problems to sort out.  As long as all logic is encapsulated in tasks, we're fine.  Additionally, the most common use of `Await` (citation needed) is via the lower-case `await` constructor, and not `awaitOr`.  If you use lower-case `await`, then the error case for `rcv` just returns `halt`, meaning that we're really only evaluating `rcv` once: when the task completes.